src/Strings.f90

```diff
@@ -364,81 +371,166 @@ FUNCTION cchar_to_fchar(cchar) RESULT(fchar)
-    ENDIF
+      stt = stt + sepLoc + LEN(separator) - 1
+      sepLoc = INDEX(this%s(stt:stp),separator)
+    ENDDO
+    !Account for strings that don't end in a delimiter
+    IF(stt <= stp) nSplits = nSplits + 1
     ALLOCATE(sub_str(nSplits))
 
-    ! Split along delimiters and store in the provided array
     stt = 1
     iSplit = 0
-    DO WHILE(INDEX(this%s(stt:stp),separator) > 0)
-      IF(INDEX(this%s(stt:stp),separator) > 1) THEN
-        iSplit = iSplit + 1
-        ! Strip out the string...subtract 2 (1 for exclusive, 1 for delimiter)
-        sub_str(iSplit) = &
-            this%s(stt:stt+INDEX(this%s(stt:stp),separator) - 2)
-        stt = stt + INDEX(this%s(stt:stp),separator) + LEN(separator) - 1
-      ELSE
-        stt = stt + LEN(separator)
-      ENDIF
+    sepLoc = INDEX(this%s(stt:stp),separator)
+    DO WHILE(sepLoc > 0)
+      iSplit = iSplit + 1
+      sub_str(iSplit) = this%s(stt:stt+sepLoc-2)
+      stt = stt + sepLoc + LEN(separator) - 1
+      sepLoc = INDEX(this%s(stt:stp),separator)
     ENDDO
     ! If the string ends with a word, then it wasn't snatched out before
-    IF(.NOT.(this%s(stp-LEN(separator)+1:stp) == separator)) THEN
+    IF(iSplit < nSplits) THEN
       sub_str(nSplits) = &
           this%s(INDEX(this%s(1:stp),separator,.TRUE.)+LEN(separator):stp)
     ENDIF
   ENDFUNCTION split_string
 !
 !-------------------------------------------------------------------------------
+!> @brief returns a copy of the string with every occurrence of a specified
+!> search pattern with the designated replacement pattern. This routine will
+!> automatically account for size changes of the underlying character array. If
+!> the search pattern is not found, then the original string is returned
+!>
+FUNCTION replace_pattern(this,oldPat,newPat) RESULT(retStr)
+  CLASS(StringType),INTENT(IN) :: this
+  CHARACTER(LEN=*),INTENT(IN) :: oldPat
+  CHARACTER(LEN=*),INTENT(IN) :: newPat
+  CHARACTER(LEN=:),ALLOCATABLE :: retStr
+  !
+  TYPE(StringType),ALLOCATABLE :: tokens(:)
+  INTEGER(SIK) :: iToken
+
+  !Split out the parts to be retained
+  tokens = this%split(oldPat)
+
+  !Set-up the new string
+  IF(SIZE(tokens) > 1) THEN
+    retStr = ''
+  ELSEIF(SIZE(tokens) == 1) THEN
+    retStr = char(tokens(1))
+    RETURN
+  ELSE
+    retStr = ''
+    RETURN
+  ENDIF
+
+  DO iToken=1,SIZE(tokens)-1
+    IF(tokens(iToken) == '') THEN
+      retStr = retStr//newPat
+    ELSE
+      retStr = retStr//tokens(iToken)//newPat
+    ENDIF
+  ENDDO
+  IF(tokens(SIZE(tokens)) /= '') THEN
+    retStr = retStr//tokens(SIZE(tokens))
+  ELSE
+    retStr = retStr//newPat
+  ENDIF
+
+ENDFUNCTION replace_pattern
+!
+!-------------------------------------------------------------------------------
 !> @brief replaces a slice from a string with the designated character(s). If
 !> the replacement characters are not the same size as the slice, the characters
 !> are repeated until the slice is filled, any excess is truncated
```

Comment:
So this function was born out a few loops I had noticed getting repeated, typically for "overwriting" stuff with spaces. So I tried to generalize it a bit, but the use case you mention is not really covered by either implementation. I don't think it would be an unreasonable addition, just not sure it is anything we currently use.

Reply:
Ok.  I'm fine with this implementation if that's what's needed.  Do you happen to know if `replace` is a common method in other languages?

Reply:
Python implements one and I think Java does too. Both as far as I know replace one search pattern with a new pattern. Neither implement a slice version

Reply:
I looked at the python implementation and it does resize the string if the pattern is a different length from the replaced pattern.

However, we don't NEED to match the python version if this is what's useful to us.  The behavior just struck me as slightly different than I would have expected.